File: Haskell_Solutions/3.hs
main :: IO()
main = do
    putStrLn$show$last$gen_prime_factors(600851475143)
    --putStrLn$show$fromInteger$toInteger$5

gen_prime_factors :: Int -> [Int]
gen_prime_factors n = [x | x <- [2,3..round$sqrt(ub)], (n `mod` x) == 0, is_prime x (x - 1)]
    where ub = fromIntegral$n

is_prime :: Int -> Int -> Bool
is_prime n m
        | m == 1 = True
        | n `mod` m == 0 = False
        | otherwise = is_prime n (m - 1)

File: Python_Solutions/3.py
import numpy as np


def is_prime(p):
    for d_ in range(3, int(np.sqrt(p))):
        if p % d_ == 0:
            return False
    return True


n = 600851475143
largest_prime = -1
for d in range(2, int(np.sqrt(n))):
    if n % d == 0 and is_prime(d):
        largest_prime = d
print(largest_prime)
